hal/pinmap.h

```diff
@@ -42,6 +48,153 @@ void     pinmap_pinout(PinName pin, const PinMap *map);
+ *         spi_master_miso_pinmap(),
+ *         spi_master_clk_pinmap(),
+ *         spi_master_cs_pinmap()
+ *     };
+ *     PinName *pins[] = {
+ *         &mosi,
+ *         &miso,
+ *         &sclk,
+ *         &ssel
+ *     };
+ *     if (pinmap_find_peripheral_pins(pins_ff, pins_avoid, per, maps, pins, sizeof(maps) / sizeof(maps[0]))) {
+ *         printf("Found SPI pins to test instance %i with:\n"
+ *                "  mosi=%s\n"
+ *                "  miso=%s\n"
+ *                "  sclk=%s\n"
+ *                "  ssel=%s\n", per,
+ *                pinmap_ff_default_pin_to_string(mosi),
+ *                pinmap_ff_default_pin_to_string(miso),
+ *                pinmap_ff_default_pin_to_string(sclk),
+ *                pinmap_ff_default_pin_to_string(ssel));
+ *     } else {
+ *         printf("Could not find SPI combination to test %i\n", per);
+ *     }
+ *     return 0;
+ * }
+ * @endcode
+ *
+ * @param whitelist List of pins to choose from
+ * @param blacklist List of pins which cannot be used
+ * @param per Peripheral to which the pins belong
+ * @param maps An array of pin maps to select from
+ * @param pins An array of pins to find. Pins already set to a value will be
+ *             left unchanged. Only pins initialized to NC will be updated by this function
+ * @param count The size of maps and pins
+ * @return true if a suitable combination of pins was found and
+ *         written to the pins array, otherwise false
+ */
+bool pinmap_find_peripheral_pins(const PinList *whitelist, const PinList *blacklist, int per, const PinMap *const *maps, PinName **pins, uint32_t count);
+
+/**
+ * Check if the pin is in the list
+ *
+ * @param list pin list to check
+ * @param pin pin to check for in the list
+ * @return true if the pin is in the list, false otherwise
+ */
+bool pinmap_list_has_pin(const PinList *list, PinName pin);
+
+/**
+ * Get the pin list of pins to avoid during testing
+ *
+ * The restricted pin list is used to indicate to testing
+ * that a pin should be skipped due to some caveat about it.
+ * For example, using USBRX and USBTX during tests will interfere
+ * with the test runner and should be avoided.
+ *
+ * Targets should override the weak implementation of this
+ * function if they have additional pins which should be
+ * skipped during testing.
+ *
+ * @return Pointer to a pin list of pins to avoid
+ */
+const PinList *pinmap_restricted_pins(void);
+
+#ifdef TARGET_FF_ARDUINO
+
+/**
+ * Get the pin list of the Arduino form factor
+ *
+ * @return Pointer to the Arduino pin list
+ */
+const PinList *pinmap_ff_arduino_pins(void);
+
+/**
+ * Get the string representation of a form factor pin
+ *
+ * @param pin Pin to get a string for
+ * @return String representing the form factor pin
+ */
+const char *pinmap_ff_arduino_pin_to_string(PinName pin);
+
+/* Default to arduino form factor if unspecified */
+#ifndef MBED_CONF_TARGET_DEFAULT_FORM_FACTOR
+#define MBED_CONF_TARGET_DEFAULT_FORM_FACTOR arduino
```

Comment:
As there is a single "default" form factor there, how would we go about testing say ST boards which have both Arduino and Morpho headers?

Reply:
The default would be the one tested as part of CI. Multiple form factors can be specified in targets.json but only one can be tested as part of CI.

Reply:
OK, I guess it wouldn't be too hard to amend this in the future if needed.
In the case of ST, @jeromecoutant which form factor is higher priority?